In [1]:
import pandas as pd 
import scipy
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
 

In [6]:
train_df = pd.read_csv("./Dataset/train.csv")
test_df = pd.read_csv("./Dataset/test.csv")

In [7]:
train_df_c = train_df.copy()
test_df_c = train_df.copy()

In [4]:
train_df_c.head()

Student ID   Age Gender   Home Region  Home City  \
0  4f14c50d-162e-4a15-9cf0-ec129c33bcf0  37.0    ذكر  منطقة الرياض     الرياض   
1  0599d409-876b-41a5-af05-749ef0e77d32  21.0    ذكر    منطقة عسير  خميس مشيط   
2  38a11c0e-4afc-4261-9c64-e94cc0a272fb  24.0    ذكر  منطقة الرياض     الرياض   
3  1693e85b-f80e-40ce-846f-395ddcece6d3  23.0    ذكر  منطقة الرياض     الرياض   
4  98a0e8d0-5f80-4634-afd8-322aa0902863  23.0    ذكر  منطقة الرياض     الرياض   

                             Program ID Program Main Category Code  \
0  453686d8-4023-4506-b2df-fac8b059ac26                       PCRF   
1  cc8e4e42-65d5-4fa1-82f9-6c6c2d508b60                       APMR   
2  e006900d-05a9-4c2b-a36f-0ffb9fce44cd                       APMR   
3  2ec15f6b-233b-428a-b9f5-e40bc8d14cf9                       TOSL   
4  d32da0e9-1aed-48c3-992d-a22f9ccc741e                       CAUF   

  Program Sub Category Code Technology Type Program Skill Level  ...  \
0                      PCRF             NaN                 NaN  ...   
1                      SWPS             NaN               متوسط  ...   
2                       NaN             NaN               متوسط  ...   
3                      TOSL             NaN                 NaN  ...   
4                      SWPS         تقليدية               متوسط  ...   

  Completed Degree Level of Education                Education Speaciality  \
0              نعم        البكالوريوس                     هندسة حاسب الالي   
1              نعم        البكالوريوس           الإذاعة والتلفزيون والفيلم   
2              نعم        البكالوريوس               Information Technology   
3              نعم        البكالوريوس  حوسبة تطبيقية - (مسار شبكات الحاسب)   
4               لا        البكالوريوس              نظم المعلومات الحاسوبية   

                          College University Degree Score  \
0                             NaN                    2.44   
1        الفنون والعلوم الإنسانية                    5.00   
2                             NaN                    3.50   
3                             NaN                    3.55   
4  تكنولوجيا الاتصالات والمعلومات                    4.00   

  University Degree Score System Employment Status Job Type  Still Working  Y  
0                            4.0          غير موظف      NaN            NaN  0  
1                            5.0              طالب      NaN            NaN  0  
2                            5.0              موظف      NaN            NaN  0  
3                            5.0              خريج      NaN            NaN  0  
4                            5.0               NaN      NaN            NaN  0  

[5 rows x 24 columns]

# Invistaging if there's any duplicates 

In [10]:
def check_dublicate(train,test):
    """
    check wither theres a 'leak' in the test set from the train set

    Args:
        train (_type_): dataframe
        test (_type_): dataframe
        
    Return: 
        the leaked record or the false  
    """
    hash_train = set(train)
    hash_test = set(test)
    
    leaks = list(hash_train.intersection(hash_test))
    has_leak = len(leaks) > 0
    print(len(leaks))
    return has_leak, leaks    
    
has_leak, leaks = check_dublicate(train_df_c['Student ID'],test_df_c['Student ID'])
print(has_leak,leaks)

5196
True ['1ad6b961-5e47-4985-ba40-547f9ecef363', 'c962d6fd-c75e-4c28-8e38-83a90ed94b4e', '9462fc01-3ca1-4fbb-ba80-5b2deb186368', 'fec28466-5993-4812-b94e-03b39bce8e38', 'eb1d932c-99e2-4e51-9057-29e572476a70', 'd42b9353-3354-4f0a-895b-efb6c51e7369', '264c0985-2c31-4f5b-ae79-ed320d967c2d', '4a43ac72-31bb-4faf-907c-fa7076f68e18', '592ef062-384c-42ad-a482-6adfb7d8f930', '687109e2-b0a5-4769-ba86-f8a30759af33', 'c2479683-2b31-4e65-873d-bf05414c31cb', 'a4be6795-4575-40a8-9787-ce34c2caa8be', 'dc082384-ffab-4170-a70a-25f9b83ce834', '9738a10e-00cd-47f6-a067-7324ddb709b5', 'b160b0f4-9d46-4acc-beac-58eee84f5338', '58eb4677-ece5-449b-9d43-879b57c554ed', 'becdeb46-b0cf-4333-a0f2-d2d284923438', 'bd145dcf-5689-4750-9ab5-f8d63564a7f8', '8518b780-4161-46c9-831e-32475fd7d16f', '6b6285bc-dc1a-43c9-b870-cc79742db22e', 'f3eb4fdf-a1cc-43f8-b1ce-2761c141c519', '22ee8044-66f4-4e9e-85bf-ba92b9785e76', '781f94ee-3017-4d62-a396-6cfa0d050f36', 'fb40b25c-6dd9-4512-aa28-ebb0b0eba1d4', 'bdc27e03-6b20-4740-8a82-c6d9

In [14]:
train_df_c.info

<bound method DataFrame.info of                                 Student ID   Age Gender   Home Region  \
0     4f14c50d-162e-4a15-9cf0-ec129c33bcf0  37.0    ذكر  منطقة الرياض   
1     0599d409-876b-41a5-af05-749ef0e77d32  21.0    ذكر    منطقة عسير   
2     38a11c0e-4afc-4261-9c64-e94cc0a272fb  24.0    ذكر  منطقة الرياض   
3     1693e85b-f80e-40ce-846f-395ddcece6d3  23.0    ذكر  منطقة الرياض   
4     98a0e8d0-5f80-4634-afd8-322aa0902863  23.0    ذكر  منطقة الرياض   
...                                    ...   ...    ...           ...   
6543  cd196579-9590-441b-8787-41078f3cee25  31.0   أنثى  منطقة الرياض   
6544  37bfc11c-ff8c-42dc-9cf9-0d13bb8f7131  27.0   أنثى  منطقة القصيم   
6545  fc114302-a79f-439f-a08b-fe0a51cf839e  24.0   أنثى  منطقة الرياض   
6546  4b6d9a36-4402-4c75-bc3a-fca927dbaf65  25.0    ذكر  منطقة الرياض   
6547  008f3386-0d43-45a4-8372-b282e5a0101a  37.0   أنثى  منطقة الرياض   

      Home City                            Program ID  \
0        الرياض  453686d8-4023-450

In [12]:
test_df_c[test_df_c['Student ID'] == '1ad6b961-5e47-4985-ba40-547f9ecef363']

Student ID   Age Gender   Home Region  \
482   1ad6b961-5e47-4985-ba40-547f9ecef363  25.0    ذكر  منطقة الرياض   
3178  1ad6b961-5e47-4985-ba40-547f9ecef363  25.0    ذكر  منطقة الرياض   

     Home City                            Program ID  \
482     الرياض  2ef39820-72d1-4d8f-be97-2d71f3737685   
3178    الرياض  55a8f0c7-7fdc-4574-9828-44e5979f7c78   

     Program Main Category Code Program Sub Category Code Technology Type  \
482                        PCRF                      PCRF         تقليدية   
3178                       PCRF                      PCRF         تقليدية   

     Program Skill Level  ... Completed Degree Level of Education  \
482                متقدم  ...              نعم        البكالوريوس   
3178               متقدم  ...              نعم        البكالوريوس   

         Education Speaciality  College University Degree Score  \
482   حاسب الي - نظم المعلومات      NaN                    4.45   
3178  حاسب الي - نظم المعلومات      NaN                    4.45   

     University Degree Score System Employment Status Job Type  Still Working  \
482                             5.0          غير موظف      NaN            NaN   
3178                            5.0          غير موظف      NaN            NaN   

      Y  
482   0  
3178  0  

[2 rows x 24 columns]

In [13]:
train_df[train_df['Student ID'] == '1ad6b961-5e47-4985-ba40-547f9ecef363']

Student ID   Age Gender   Home Region  \
482   1ad6b961-5e47-4985-ba40-547f9ecef363  25.0    ذكر  منطقة الرياض   
3178  1ad6b961-5e47-4985-ba40-547f9ecef363  25.0    ذكر  منطقة الرياض   

     Home City                            Program ID  \
482     الرياض  2ef39820-72d1-4d8f-be97-2d71f3737685   
3178    الرياض  55a8f0c7-7fdc-4574-9828-44e5979f7c78   

     Program Main Category Code Program Sub Category Code Technology Type  \
482                        PCRF                      PCRF         تقليدية   
3178                       PCRF                      PCRF         تقليدية   

     Program Skill Level  ... Completed Degree Level of Education  \
482                متقدم  ...              نعم        البكالوريوس   
3178               متقدم  ...              نعم        البكالوريوس   

         Education Speaciality  College University Degree Score  \
482   حاسب الي - نظم المعلومات      NaN                    4.45   
3178  حاسب الي - نظم المعلومات      NaN                    4.45   

     University Degree Score System Employment Status Job Type  Still Working  \
482                             5.0          غير موظف      NaN            NaN   
3178                            5.0          غير موظف      NaN            NaN   

      Y  
482   0  
3178  0  

[2 rows x 24 columns]

Insights
- yes there's a leak and it's not normal the leak is 5196 record out of 6548
- it might be ***usefull*** to copy the solution of (get inspired by it) https://www.kaggle.com/competitions/arabic-poem-classification/discussion/483099